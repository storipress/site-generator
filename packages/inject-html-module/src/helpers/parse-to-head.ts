import type { HTMLElement, Node } from 'node-html-parser'
import { NodeType, parse } from 'node-html-parser'
import type { Head } from 'zhead'

export type ParsedHead = Required<Pick<Head, 'link' | 'meta' | 'style' | 'script' | 'noscript'>>

export interface ParsedInjectHTML {
  head: ParsedHead
  html: string
}

const SUPPORTED_TAGS = ['script', 'link', 'meta', 'style', 'noscript'] as const
const SUPPORTED_TAG_SET: Set<string> = new Set(SUPPORTED_TAGS)

export function parseToInjectHTML(html: string, tagPosition: 'head' | 'bodyClose' = 'head'): ParsedInjectHTML {
  const parsedHead: ParsedHead = { script: [], link: [], meta: [], style: [], noscript: [] }
  let otherHTML = ''
  const root = parse(html)

  for (const node of root.childNodes) {
    if (isHtmlElement(node)) {
      const tagName = getTagName(node)
      if (isSupportTag(tagName)) {
        parsedHead[tagName].push({
          ...node.attributes,
          ...(node.innerHTML ? { innerHTML: node.innerHTML } : {}),
          // @ts-expect-error no type
          tagPriority: tagPosition === 'head' ? 'high' : 'low',
          tagPosition,
        })
      } else {
        otherHTML += node.outerHTML
      }
    } else {
      otherHTML += node.rawText
    }
  }

  return {
    head: parsedHead,
    html: otherHTML,
  }
}

export function mergeParsedHTML(heads: ParsedInjectHTML[]): ParsedInjectHTML {
  const mergedHead: ParsedHead = { script: [], link: [], meta: [], style: [], noscript: [] }
  let otherHTML = ''

  for (const head of heads) {
    for (const tag of SUPPORTED_TAGS) {
      mergedHead[tag].push(...head.head[tag])
    }
    otherHTML += head.html
  }

  return {
    head: mergedHead,
    html: otherHTML,
  }
}

function isSupportTag(tag: string): tag is (typeof SUPPORTED_TAGS)[number] {
  return SUPPORTED_TAG_SET.has(tag.toLowerCase())
}

function isHtmlElement(el: Node): el is HTMLElement {
  return el.nodeType === NodeType.ELEMENT_NODE
}

function getTagName(el: HTMLElement): string {
  return el.tagName.toLowerCase()
}
