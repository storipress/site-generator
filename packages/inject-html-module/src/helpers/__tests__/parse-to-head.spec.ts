import { expect, it } from 'vitest'
import { parseToInjectHTML } from '../parse-to-head'

it.each([
  '<meta name="google-site-verification" content="kYgLy4ZCjuR-ZxaISEs0ofzt5wpJAd9nTZ51hYdC_zs" />',
  '<meta name="google-site-verification" content="kYgLy4ZCjuR-ZxaISEs0ofzt5wpJAd9nTZ51hYdC_zs" >',
  '<meta name="google-site-verification" content="kYgLy4ZCjuR-ZxaISEs0ofzt5wpJAd9nTZ51hYdC_zs">',
])('parse google search console meta', (html: string) => {
  expect(parseToInjectHTML(html)).toMatchSnapshot()
})

it.each(['<>', '</>', '<   >'])('should correctly handle for invalid html `%s`', (html: string) => {
  expect(parseToInjectHTML(html)).toMatchSnapshot()
})
