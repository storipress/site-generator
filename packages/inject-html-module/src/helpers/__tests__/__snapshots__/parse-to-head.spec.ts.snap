// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`parse google search console meta 1`] = `
{
  "head": {
    "link": [],
    "meta": [
      {
        "content": "kYgLy4ZCjuR-ZxaISEs0ofzt5wpJAd9nTZ51hYdC_zs",
        "name": "google-site-verification",
        "tagPosition": "head",
        "tagPriority": "high",
      },
    ],
    "noscript": [],
    "script": [],
    "style": [],
  },
  "html": "",
}
`;

exports[`parse google search console meta 2`] = `
{
  "head": {
    "link": [],
    "meta": [
      {
        "content": "kYgLy4ZCjuR-ZxaISEs0ofzt5wpJAd9nTZ51hYdC_zs",
        "name": "google-site-verification",
        "tagPosition": "head",
        "tagPriority": "high",
      },
    ],
    "noscript": [],
    "script": [],
    "style": [],
  },
  "html": "",
}
`;

exports[`parse google search console meta 3`] = `
{
  "head": {
    "link": [],
    "meta": [
      {
        "content": "kYgLy4ZCjuR-ZxaISEs0ofzt5wpJAd9nTZ51hYdC_zs",
        "name": "google-site-verification",
        "tagPosition": "head",
        "tagPriority": "high",
      },
    ],
    "noscript": [],
    "script": [],
    "style": [],
  },
  "html": "",
}
`;

exports[`should correctly handle for invalid html \`<   >\` 1`] = `
{
  "head": {
    "link": [],
    "meta": [],
    "noscript": [],
    "script": [],
    "style": [],
  },
  "html": "<   >",
}
`;

exports[`should correctly handle for invalid html \`</>\` 1`] = `
{
  "head": {
    "link": [],
    "meta": [],
    "noscript": [],
    "script": [],
    "style": [],
  },
  "html": "</>",
}
`;

exports[`should correctly handle for invalid html \`<>\` 1`] = `
{
  "head": {
    "link": [],
    "meta": [],
    "noscript": [],
    "script": [],
    "style": [],
  },
  "html": "<>",
}
`;
